### PR TITLE
containerd integration: imageservice: add contexts to various methods (step 1)

### PR DESCRIPTION
**Pull request by thaJeztah**
- depends on https://github.com/moby/moby/pull/44091
- Splitting https://github.com/moby/moby/pull/43828 into smaller bits to see where the failures in that PR come from.

### CODE REVIEW

**Comment by thaJeztah** [comment]
whoop; at least this one is green; I updated https://github.com/moby/moby/pull/43818 based on this PR (to start with the minimum amount of changes for that PR to work)

**Review by rumpl — APPROVED** [review]

**Review by corhere — CHANGES_REQUESTED** [review]
Plumbing contexts through a function inevitably makes it possible for an operation to fail for a completely benign reason: the context was canceled. A function could misbehave if it is not prepared for the possibility that context-accepting function calls could be canceled. Best-case scenario it returns an inaccurate error if a context is canceled. Worst-case scenario is a timing-dependent bug which occurs only if the context is canceled at a specific point in the operation's execution. Given the risk of introducing bugs which could be really troublesome to debug—or consistently reproduce—I think we have to make sure all functions are cancelation-aware when we plumb contexts through them. Since Go convention is to signal that a function call was canceled by returning the `ctx.Err()` value, cancelation-awareness is necessarily an aspect of error handling. So I think we have to plumb error-handling up into the caller at the same time as we plumb contexts down into a callee function.

**Comment by thaJeztah** [comment]
^^ also discussing this in the maintainers meeting; @corhere had some ideas on different approach on handling the context, and context-cancellation, which may be a good topic for our containerd-integration call ./cc @bsousaa @rumpl 

I think next meeting is on a "whaleness" day, but perhaps it would be an idea to write down some approaches so that we can decide which one best fits our situation best.

**Comment by corhere** [comment]
A function must be prepared to handle the cancelation of any function calls it passes a (potentially) cancelable context into. As a corollary, functions for which it can be statically proven that under no circumstances can a cancelable context be passed into a function call do not need to be cancelation-aware. Therefore:
- `func Foo(_ context.Context)` generally does not need to be cancelation-aware as it does not propagate its context argument into any function calls, and
- a function call which passes in a statically provably not-cancelable context (`context.Background()` or `context.TODO()`) cannot be canceled.

I think it is possible to to plumb contexts through iteratively without getting overwhelmed or missing anything. The general idea is to always maintain some invariants while refactoring which do not require having to recursively audit all the callers or callees of a refactored function. Here's what I have come up with:
1. A cancelable function must signal to its caller when its operation is canceled by returning (an error which wraps) `ctx.Err()`.
2. Any function which accepts a context argument and returns an error value must be assumed to be a cancelable function. (It could be refactored to become cancelable at any time, if it is not already.) Any call sites to that function must be cancelation-aware unless it can be statically proved that the context argument passed in is not cancelable. In most cases, cancelation-awareness implies that the calling function must itself be cancelable.

Refactoring could proceed iteratively, touching a chosen subset of the codebase at a time while maintaining the invariants. When a context argument is added to a function, all its call sites need only be modified to pass in `context.TODO()`, without needing to worry about cancelation-awareness. Reviewers can skim the call sites and focus on auditing only the function being refactored. When a context argument needs to be added to an interface method and the implementations need to be updated to match, the implementation functions' signatures can be modified by discarding the argument value (e.g. `func (*Foo) Do(_ context.Context)`) so that the function bodies do not need to be touched. Any gaps in context propagation can be found and iteratively refactored by searching for and replacing any `context.TODO()` calls.

The major flaw in the iterative approach is that values cannot be propagated across the `context.TODO()` gaps. There are ways to "derive" a context which propagates values but not deadlines or cancelation so that values can be propagated through cancelation-unaware function calls. [The trivial but wrong way](https://github.com/golang/go/wiki/CodeReviewComments#contexts:~:text=Don%27t%20create%20custom%20Context%20types) would be to create a custom Context type which stubs out the `Deadline`, `Done`, and `Err` methods. The right way would be to start a new context chain from `context.Background()` and explicitly copy the desired values into it, assuming you have access to the necessary context keys.

**Comment by thaJeztah** [comment]
@rumpl looks like this may be related 🤔 

```
=== FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform/different_os (0.00s)
    create_test.go:484: assertion failed: expression is false: client.IsErrNotFound(err): Error response from daemon: namespace is required: failed precondition
    --- FAIL: TestCreateDifferentPlatform/different_os (0.00s)

=== FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform/different_cpu_arch (0.00s)
    create_test.go:493: assertion failed: expression is false: client.IsErrNotFound(err): Error response from daemon: namespace is required: failed precondition
    --- FAIL: TestCreateDifferentPlatform/different_cpu_arch (0.00s)

=== FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform (0.02s)
```

File: daemon/cluster/executor/container/adapter.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func (c *containerAdapter) pullImage(ctx context.Context) error {
 	named, err := reference.ParseNormalizedNamed(spec.Image)
 	if err == nil {
 		if _, ok := named.(reference.Canonical); ok {
-			_, err := c.imageBackend.GetImage(spec.Image, imagetypes.GetImageOpts{})
+			_, err := c.imageBackend.GetImage(ctx, spec.Image, imagetypes.GetImageOpts{})
 			if err == nil {
 				return nil
 			}
```
Comment (corhere) [inline daemon/cluster/executor/container/adapter.go:82]:
```suggestion
			_, err := c.imageBackend.GetImage(ctx, spec.Image, imagetypes.GetImageOpts{})
			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
				return err
			}
			if err == nil {
				return nil
			}
```

File: daemon/images/image_builder.go [hunk 2]
```diff
@@ -212,7 +212,7 @@ func (i *ImageService) GetImageAndReleasableLayer(ctx context.Context, refOrID s
 	}
 
 	if opts.PullOption != backend.PullOptionForcePull {
-		img, err := i.GetImage(refOrID, imagetypes.GetImageOpts{Platform: opts.Platform})
+		img, err := i.GetImage(ctx, refOrID, imagetypes.GetImageOpts{Platform: opts.Platform})
 		if err != nil && opts.PullOption == backend.PullOptionNoPull {
```
Comment (corhere) [inline daemon/images/image_builder.go:216]: That TODO comment below the fold should probably be addressed now that `err` could be `context.Canceled` or `context.DeadlineExceeded`.
https://github.com/moby/moby/blob/0db50996b76fa009b4e905dff35be3a1d308f698/daemon/images/image_builder.go#L219

File: daemon/images/image_pull.go [hunk 3]
```diff
@@ -64,7 +64,7 @@ func (i *ImageService) PullImage(ctx context.Context, image, tag string, platfor
 		// we allow the image to have a non-matching architecture. The code
 		// below checks for this situation, and returns a warning to the client,
 		// as well as logging it to the daemon logs.
-		img, err := i.GetImage(image, imagetypes.GetImageOpts{Platform: platform})
+		img, err := i.GetImage(ctx, image, imagetypes.GetImageOpts{Platform: platform})
```
Comment (corhere) [inline daemon/images/image_pull.go:67]: `err` needs to be returned when that special case below does not apply.

File: daemon/images/image_builder.go (hunk 2 above)
Comment (rumpl) [reply daemon/images/image_builder.go:216]: Done

File: daemon/images/image_pull.go (hunk 3 above)
Comment (rumpl) [reply daemon/images/image_pull.go:67]: Done 👍

**Comment by thaJeztah** [comment]
> @rumpl looks like this may be related 🤔
> 
> ```
> === FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform/different_os (0.00s)
>     create_test.go:484: assertion failed: expression is false: client.IsErrNotFound(err): Error response from daemon: namespace is required: failed precondition
>     --- FAIL: TestCreateDifferentPlatform/different_os (0.00s)
> 
> === FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform/different_cpu_arch (0.00s)
>     create_test.go:493: assertion failed: expression is false: client.IsErrNotFound(err): Error response from daemon: namespace is required: failed precondition
>     --- FAIL: TestCreateDifferentPlatform/different_cpu_arch (0.00s)
> 
> === FAIL: github.com/docker/docker/integration/container TestCreateDifferentPlatform (0.02s)
> ```

This one looks to be failing still (on `windows-2022 / run / integration-test (builtin, ./...)`). Not 100% sure what's happening there; it's a bit weird that it's failing on that one, because that's `builtin`, so should even use containerd? Error is coming from https://github.com/moby/moby/blob/067062129171dba592638e8e3a0efbfa8a71376b/vendor/github.com/containerd/containerd/namespaces/context.go#L72

File: daemon/images/image_pull.go [hunk 4]
```diff
@@ -73,6 +73,9 @@ func (i *ImageService) PullImage(ctx context.Context, image, tag string, platfor
 			progress.Messagef(po, "", `WARNING: %s`, err.Error())
 			logrus.WithError(err).WithField("image", image).Warn("ignoring platform mismatch on single-arch image")
 		}
+		if err != nil && img == nil {
```
Comment (corhere) [inline daemon/images/image_pull.go:76]: What are the other special cases where `err != nil && img != nil`? Why is silently ignoring the error the correct behaviour?

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/44091 was merged, so this PR can be rebased (to drop the first commit)

File: daemon/images/image_pull.go (hunk 4 above)
Comment (rumpl) [reply daemon/images/image_pull.go:76]: See the comment above:
```
// Note that this is a special case where GetImage returns both an image
// and an error: https://github.com/docker/docker/blob/v20.10.7/daemon/images/image.go#L175-L183
```

**Comment by rumpl** [comment]
> #44091 was merged, so this PR can be rebased (to drop the first commit)

Done

File: daemon/images/image_pull.go (hunk 4 above)
Comment (rumpl) [reply daemon/images/image_pull.go]: Right, fixed

File: daemon/create.go [hunk 5]
```diff
@@ -69,7 +71,11 @@ func (daemon *Daemon) containerCreate(opts createOpts) (containertypes.CreateRes
...
 
 	if opts.params.Platform == nil && opts.params.Config.Image != "" {
-		if img, _ := daemon.imageService.GetImage(opts.params.Config.Image, imagetypes.GetImageOpts{Platform: opts.params.Platform}); img != nil {
+		img, err := daemon.imageService.GetImage(ctx, opts.params.Config.Image, imagetypes.GetImageOpts{Platform: opts.params.Platform})
+		if err != nil {
+			return containertypes.CreateResponse{}, err
+		}
+		if img != nil {
 			p := maximumSpec()
 			imgPlat := v1.Platform{
```
Comment (thaJeztah) [inline daemon/create.go:80]: Not for this PR, but (also related to the other question about `nil` error and `nil` image), perhaps we should look at `GetImage` to return a sentinel error in some of these conditions.

**Review by thaJeztah — COMMENTED** [review]
LGTM

File: daemon/images/image_pull.go [hunk 6]
```diff
@@ -64,14 +64,18 @@ func (i *ImageService) PullImage(ctx context.Context, image, tag string, platfor
...
 		// and an error: https://github.com/docker/docker/blob/v20.10.7/daemon/images/image.go#L175-L183
 		if errdefs.IsNotFound(err) && img != nil {
 			po := streamformatter.NewJSONProgressOutput(outStream, false)
 			progress.Messagef(po, "", `WARNING: %s`, err.Error())
 			logrus.WithError(err).WithField("image", image).Warn("ignoring platform mismatch on single-arch image")
+		} else {
+			if err != nil {
+				return err
+			}
 		}
```
Comment (corhere) [inline daemon/images/image_pull.go:79]: Nit:
```suggestion
		} else if err != nil {
			return err
		}
```

**Review by corhere — APPROVED** [review]
LGTM! :shipit:

File: daemon/images/image_pull.go (hunk 6 above)
Comment (thaJeztah) [reply daemon/images/image_pull.go]: LOL... I _wanted_ to make that comment, and thought "let it be" 😂

Comment (thaJeztah) [reply daemon/images/image_pull.go]: I want to enable more linters; looks like a bunch were deprecated, and there's a couple of interesting new ones; was testing some of them already on our code.

Not sure if any of them picked up this one though.

Comment (rumpl) [reply daemon/images/image_pull.go]: I’m very good at dodging linters

**Pull request merged**